### ISSUE TITLE
Unknown method: DisplayPinCode is not a valid method of interface org.bluez.Agent1

### ISSUE THREAD

**Issue by patrakov** [blueman-project/blueman#120]
Blueman from git does not allow me to use a bluetooth keyboard. In bluetoothd logs, I see this:

```
aep-haswell bluetoothd[655]: Agent replied with an error: org.freedesktop.DBus.Error.UnknownMethod, Traceback (most recent call last):
    File "/usr/lib64/python2.7/site-packages/dbus/service.py", line 654, in _message_cb
        (candidate_method, parent_method) = _method_lookup(self, method_name, interface_name)
    File "/usr/lib64/python2.7/site-packages/dbus/service.py", line 246, in _method_lookup
        raise UnknownMethodException('%s is not a valid method of interface %s' % (method_name, dbus_interface))
    UnknownMethodException: org.freedesktop.DBus.Error.UnknownMethod: Unknown method: DisplayPinCode is not a valid method of interface org.bluez.Agent1
```

**Comment by cschramm** [comment]
Can you test the displaypincode branch?

You've actually stumbled upon to separate issues. First, DisplayPinCode was never implemented in BlueZ. The patch from the branch fixes that. Second, while the similar DisplayPasskey method was already implemented, it isn't very useful. It will only show the passkey in the terminal output; as will the new DisplayPinCode method. If that works for you, I will implement an actual desktop notification for both to enable this type of pairing.

**Comment by patrakov** [comment]
I have just tested this branch with that keyboard. Indeed, it prints a pincode that, when typed, leads to the HID authentication popup. If I authorize the device, it works. However, in the terminal output, there is a traceback.

Here is the full log from the applet. There were two pincode prompts, please ignore the first one (237058), because I was not quick enough to recognize it in the log, and attempted to type some garbage instead. The second one (552803) was typed correctly.

```
aep@aep-haswell ~ $ blueman-applet
Loading configuration plugins

** (blueman-applet:1185): WARNING **: Can't load fallback CSS resource: Failed to import: Ресурс из «/org/gnome/adwaita/gtk-fallback.css» не существует

** (blueman-applet:1185): WARNING **: Can't load fallback CSS resource: Failed to import: Ресурс из «/org/gnome/adwaita/gtk-fallback.css» не существует
blueman-applet version 1.99.alpha1 starting
Stale PID, overwriting
Using file config backend
_________
Load (/usr/lib64/python2.7/site-packages/blueman/main/PluginManager.py:54)
['AppIndicator', 'ExitItem', 'ShowConnected', 'AuthAgent', 'DhcpClient', 'NMPANSupport', 'DBusService', 'DiscvManager', 'NetUsage', 'Headset', 'KillSwitch', 'Menu', 'Networking', 'PowerManager', 'PPPSupport', 'NMDUNSupport', 'RecentConns', 'SerialManager', 'StandardItems', 'StatusIcon', 'TransferService'] 
ERROR:root:Could not find any typelib for AppIndicator3
_________
Load (/usr/lib64/python2.7/site-packages/blueman/main/PluginManager.py:54)
Unable to load plugin module AppIndicator
cannot import name AppIndicator3 
Using file config backend
_________
get_interface_version (/usr/lib64/python2.7/site-packages/blueman/bluez/BlueZInterface.py:8)
Detected BlueZ 5 
_________
__load_plugin (/usr/lib64/python2.7/site-packages/blueman/main/PluginManager.py:127)
loading <class 'blueman.plugins.applet.StatusIcon.StatusIcon'> 
_________
__load_plugin (/usr/lib64/python2.7/site-packages/blueman/main/PluginManager.py:127)
loading <class 'blueman.plugins.applet.DBusService.DBusService'> 
_________
__load_plugin (/usr/lib64/python2.7/site-packages/blueman/main/PluginManager.py:127)
loading <class 'blueman.plugins.applet.Headset.Headset'> 
Using file config backend
_________
__load_plugin (/usr/lib64/python2.7/site-packages/blueman/main/PluginManager.py:127)
loading <class 'blueman.plugins.applet.NMDUNSupport.NMDUNSupport'> 
_________
__load_plugin (/usr/lib64/python2.7/site-packages/blueman/main/PluginManager.py:127)
loading <class 'blueman.plugins.applet.SerialManager.SerialManager'> 
Using file config backend
_________
__load_plugin (/usr/lib64/python2.7/site-packages/blueman/main/PluginManager.py:127)
loading <class 'blueman.plugins.applet.Menu.Menu'> 
_________
__load_plugin (/usr/lib64/python2.7/site-packages/blueman/main/PluginManager.py:127)
loading <class 'blueman.plugins.applet.AuthAgent.AuthAgent'> 
_________
__load_plugin (/usr/lib64/python2.7/site-packages/blueman/main/PluginManager.py:127)
loading <class 'blueman.plugins.applet.ShowConnected.ShowConnected'> 
_________
__load_plugin (/usr/lib64/python2.7/site-packages/blueman/main/PluginManager.py:127)
Not loading DhcpClient because it's conflict has higher priority 
_________
__load_plugin (/usr/lib64/python2.7/site-packages/blueman/main/PluginManager.py:127)
loading <class 'blueman.plugins.applet.TransferService.TransferService'> 
_________
__load_plugin (/usr/lib64/python2.7/site-packages/blueman/main/PluginManager.py:127)
loading <class 'blueman.plugins.applet.DiscvManager.DiscvManager'> 
Using file config backend
_________
update_menuitems (/usr/lib64/python2.7/site-packages/blueman/plugins/applet/DiscvManager.py:114)
warning: Adapter is None 
_________
__load_plugin (/usr/lib64/python2.7/site-packages/blueman/main/PluginManager.py:127)
loading <class 'blueman.plugins.applet.RecentConns.RecentConns'> 
Using file config backend
_________
__load_plugin (/usr/lib64/python2.7/site-packages/blueman/main/PluginManager.py:127)
Not loading PPPSupport because it's conflict has higher priority 
_________
__load_plugin (/usr/lib64/python2.7/site-packages/blueman/main/PluginManager.py:127)
loading <class 'blueman.plugins.applet.StandardItems.StandardItems'> 
_________
__load_plugin (/usr/lib64/python2.7/site-packages/blueman/main/PluginManager.py:127)
loading <class 'blueman.plugins.applet.NMPANSupport.NMPANSupport'> 
_________
on_obex_owner_changed (/usr/lib64/python2.7/site-packages/blueman/plugins/applet/TransferService.py:70)
obex owner changed: :1.38 
Using file config backend
Traceback (most recent call last):
  File "/usr/lib64/python2.7/site-packages/blueman/main/applet/Transfer.py", line 78, in on_server_created
    self.start_server(pattern)
  File "/usr/lib64/python2.7/site-packages/blueman/main/applet/Transfer.py", line 58, in start_server
    raise Exception("Configured share directory %s does not exist" % self.Config.props.shared_path)
Exception: Configured share directory /home/aep/Public does not exist
Traceback (most recent call last):
  File "/usr/lib64/python2.7/site-packages/blueman/main/applet/Transfer.py", line 78, in on_server_created
    self.start_server(pattern)
  File "/usr/lib64/python2.7/site-packages/blueman/main/applet/Transfer.py", line 58, in start_server
    raise Exception("Configured share directory %s does not exist" % self.Config.props.shared_path)
Exception: Configured share directory /home/aep/Public does not exist
_________
on_dbus_name_owner_change (/usr/bin/blueman-applet:86)
org.bluez owner changed to :1.14 
_________
register_agent (/usr/lib64/python2.7/site-packages/blueman/plugins/applet/AuthAgent.py:57)
Registering agent 
_________
__init__ (/usr/lib64/python2.7/site-packages/blueman/main/Device.py:16)
caching initial properties 
_________
__del__ (/usr/lib64/python2.7/site-packages/blueman/main/Device.py:60)
deleting device None 
_________
Destroy (/usr/lib64/python2.7/site-packages/blueman/main/Device.py:98)
invalidating device None 
Exception AttributeError: "'NoneType' object has no attribute 'Signals'" in <bound method Device.__del__ of <Device object at 0x7fbe3f69e0f0 (blueman+main+Device+Device at 0x21c4520)>> ignored
_________
__del__ (/usr/lib64/python2.7/site-packages/blueman/main/Device.py:60)
deleting device None 
_________
Destroy (/usr/lib64/python2.7/site-packages/blueman/main/Device.py:98)
invalidating device None 
Exception AttributeError: "'NoneType' object has no attribute 'Signals'" in <bound method Device.__del__ of <Device object at 0x7fbe3f69e460 (blueman+main+Device+Device at 0x21c3d40)>> ignored
_________
__init__ (/usr/lib64/python2.7/site-packages/blueman/main/Device.py:16)
caching initial properties 
_________
__del__ (/usr/lib64/python2.7/site-packages/blueman/main/Device.py:60)
deleting device None 
_________
Destroy (/usr/lib64/python2.7/site-packages/blueman/main/Device.py:98)
invalidating device None 
Exception AttributeError: "'NoneType' object has no attribute 'Signals'" in <bound method Device.__del__ of <Device object at 0x7fbe3f69e690 (blueman+main+Device+Device at 0x218ed00)>> ignored
_________
initialize (/usr/lib64/python2.7/site-packages/blueman/plugins/applet/RecentConns.py:138)
rebuilding menu 
_________
enumerate_connections (/usr/lib64/python2.7/site-packages/blueman/plugins/applet/ShowConnected.py:45)
Found 0 existing connections 








Loading configuration plugins
Using file config backend

** (blueman-manager:1189): WARNING **: Can't load fallback CSS resource: Failed to import: Ресурс из «/org/gnome/adwaita/gtk-fallback.css» не существует

** (blueman-manager:1189): WARNING **: Can't load fallback CSS resource: Failed to import: Ресурс из «/org/gnome/adwaita/gtk-fallback.css» не существует
_________
Load (/usr/lib64/python2.7/site-packages/blueman/main/PluginManager.py:54)
['PulseAudioProfile', 'Services'] 
_________
__load_plugin (/usr/lib64/python2.7/site-packages/blueman/main/PluginManager.py:127)
loading <class 'blueman.plugins.manager.Services.Services'> 
_________
__load_plugin (/usr/lib64/python2.7/site-packages/blueman/main/PluginManager.py:127)
loading <class 'blueman.plugins.manager.PulseAudioProfile.PulseAudioProfile'> 
_________
pa_context_event (/usr/lib64/python2.7/site-packages/blueman/main/PulseAudioUtils.py:306)
1 
blueman-manager version 1.99.alpha1 starting
_________
pa_context_event (/usr/lib64/python2.7/site-packages/blueman/main/PulseAudioUtils.py:306)
2 
_________
on_bluez_name_owner_changed (/usr/bin/blueman-manager:105)
org.bluez owner changed to  :1.14 
Using file config backend
_________
get_interface_version (/usr/lib64/python2.7/site-packages/blueman/bluez/BlueZInterface.py:8)
Detected BlueZ 5 
_________
SetAdapter (/usr/lib64/python2.7/site-packages/blueman/gui/DeviceList.py:264)
None 
_________
on_adapter_changed (/usr/lib64/python2.7/site-packages/blueman/gui/manager/ManagerToolbar.py:74)
toolbar adapter /org/bluez/hci0 
_________
__init__ (/usr/lib64/python2.7/site-packages/blueman/main/Device.py:16)
caching initial properties 
_________
add_device (/usr/lib64/python2.7/site-packages/blueman/gui/DeviceList.py:310)
adding new device 
_________
do_cache (/usr/lib64/python2.7/site-packages/blueman/gui/DeviceList.py:489)
Caching new device 88:53:2E:C0:50:98 
_________
row_update_event (/usr/lib64/python2.7/site-packages/blueman/gui/manager/ManagerDeviceList.py:289)
row update event Fake False 
_________
row_update_event (/usr/lib64/python2.7/site-packages/blueman/gui/manager/ManagerDeviceList.py:289)
row update event Trusted 1 
_________
row_update_event (/usr/lib64/python2.7/site-packages/blueman/gui/manager/ManagerDeviceList.py:289)
row update event Paired 1 
_________
__init__ (/usr/lib64/python2.7/site-packages/blueman/main/Device.py:16)
caching initial properties 
_________
add_device (/usr/lib64/python2.7/site-packages/blueman/gui/DeviceList.py:310)
adding new device 
_________
do_cache (/usr/lib64/python2.7/site-packages/blueman/gui/DeviceList.py:489)
Caching new device 50:32:75:D7:1A:CC 
_________
row_update_event (/usr/lib64/python2.7/site-packages/blueman/gui/manager/ManagerDeviceList.py:289)
row update event Fake False 
_________
row_update_event (/usr/lib64/python2.7/site-packages/blueman/gui/manager/ManagerDeviceList.py:289)
row update event Trusted 1 
_________
row_update_event (/usr/lib64/python2.7/site-packages/blueman/gui/manager/ManagerDeviceList.py:289)
row update event Paired 1 
_________
init_services (/usr/lib64/python2.7/site-packages/blueman/main/Device.py:73)
Loading services 
_________
Generate (/usr/lib64/python2.7/site-packages/blueman/gui/manager/ManagerDeviceMenu.py:213)
AEP-SGS4 
_________
pa_context_event (/usr/lib64/python2.7/site-packages/blueman/main/PulseAudioUtils.py:306)
3 
_________
pa_context_event (/usr/lib64/python2.7/site-packages/blueman/main/PulseAudioUtils.py:306)
4 
_________
on_pa_ready (/usr/lib64/python2.7/site-packages/blueman/plugins/manager/PulseAudioProfile.py:24)
connected 
_________
<lambda> (/usr/lib64/python2.7/site-packages/blueman/main/PulseAudioUtils.py:318)
1 
_________
on_property_changed (/usr/lib64/python2.7/site-packages/blueman/gui/DeviceList.py:148)
adapter propery changed Discovering 1 
_________
on_adapter_property_changed (/usr/lib64/python2.7/site-packages/blueman/plugins/applet/DiscvManager.py:87)
prop Discovering 1 
_________
on_device_created (/usr/lib64/python2.7/site-packages/blueman/gui/DeviceList.py:246)
created /org/bluez/hci0/dev_20_73_AB_62_9D_B3 
_________
__init__ (/usr/lib64/python2.7/site-packages/blueman/main/Device.py:16)
caching initial properties 
_________
add_device (/usr/lib64/python2.7/site-packages/blueman/gui/DeviceList.py:310)
adding new device 
_________
do_cache (/usr/lib64/python2.7/site-packages/blueman/gui/DeviceList.py:489)
Caching new device 20:73:AB:62:9D:B3 
_________
row_update_event (/usr/lib64/python2.7/site-packages/blueman/gui/manager/ManagerDeviceList.py:289)
row update event Fake False 
_________
row_update_event (/usr/lib64/python2.7/site-packages/blueman/gui/manager/ManagerDeviceList.py:289)
row update event Trusted 0 
_________
row_update_event (/usr/lib64/python2.7/site-packages/blueman/gui/manager/ManagerDeviceList.py:289)
row update event Paired 0 
_________
on_device_property_changed (/usr/lib64/python2.7/site-packages/blueman/gui/DeviceList.py:156)
list: device_prop_ch RSSI -82 /org/bluez/hci0/dev_20_73_AB_62_9D_B3 () {} 
_________
row_update_event (/usr/lib64/python2.7/site-packages/blueman/gui/manager/ManagerDeviceList.py:289)
row update event RSSI -82 
opacity 144
_________
on_device_property_changed (/usr/lib64/python2.7/site-packages/blueman/gui/DeviceList.py:156)
list: device_prop_ch RSSI -74 /org/bluez/hci0/dev_20_73_AB_62_9D_B3 () {} 
_________
row_update_event (/usr/lib64/python2.7/site-packages/blueman/gui/manager/ManagerDeviceList.py:289)
row update event RSSI -74 
opacity 168
_________
on_property_changed (/usr/lib64/python2.7/site-packages/blueman/gui/DeviceList.py:148)
_________
on_adapter_property_changed (/usr/lib64/python2.7/site-packages/blueman/plugins/applet/DiscvManager.py:87)
prop Discovering 0 
adapter propery changed Discovering 0 
_________
init_services (/usr/lib64/python2.7/site-packages/blueman/main/Device.py:73)
Loading services 
_________
Generate (/usr/lib64/python2.7/site-packages/blueman/gui/manager/ManagerDeviceMenu.py:213)
Bluetooth V3.0 TouchPad Keyboard 
_________
on_device_property_changed (/usr/lib64/python2.7/site-packages/blueman/gui/DeviceList.py:156)
list: device_prop_ch Connected 1 /org/bluez/hci0/dev_20_73_AB_62_9D_B3 () {} 
_________
row_update_event (/usr/lib64/python2.7/site-packages/blueman/gui/manager/ManagerDeviceList.py:289)
row update event Connected 1 
_________
init_services (/usr/lib64/python2.7/site-packages/blueman/main/Device.py:73)
Loading services 
_________
DisplayPinCode (/usr/lib64/python2.7/site-packages/blueman/bluez/Agent.py:67)
DisplayPinCode (/org/bluez/hci0/dev_20_73_AB_62_9D_B3, 237058) 
_________
Generate (/usr/lib64/python2.7/site-packages/blueman/gui/manager/ManagerDeviceMenu.py:213)
Bluetooth V3.0 TouchPad Keyboard 
_________
monitor_power_levels (/usr/lib64/python2.7/site-packages/blueman/gui/DeviceList.py:178)
starting monitor 
_________
level_setup_event (/usr/lib64/python2.7/site-packages/blueman/gui/manager/ManagerDeviceList.py:350)
animating up 
_________
on_device_property_changed (/usr/lib64/python2.7/site-packages/blueman/gui/DeviceList.py:156)
list: device_prop_ch Connected 0 /org/bluez/hci0/dev_20_73_AB_62_9D_B3 () {} 
_________
_________
__init__ (/usr/lib64/python2.7/site-packages/blueman/main/Device.py:16)
row_update_event (/usr/lib64/python2.7/site-packages/blueman/gui/manager/ManagerDeviceList.py:289)
caching initial properties 
row update event Connected 0 
_________
__del__ (/usr/lib64/python2.7/site-packages/blueman/main/Device.py:60)
deleting device /org/bluez/hci0/dev_20_73_AB_62_9D_B3 
_________
Destroy (/usr/lib64/python2.7/site-packages/blueman/main/Device.py:98)
invalidating device /org/bluez/hci0/dev_20_73_AB_62_9D_B3 
_________
init_services (/usr/lib64/python2.7/site-packages/blueman/main/Device.py:73)
Loading services 
_________
Generate (/usr/lib64/python2.7/site-packages/blueman/gui/manager/ManagerDeviceMenu.py:213)
Bluetooth V3.0 TouchPad Keyboard 
_________
level_setup_event (/usr/lib64/python2.7/site-packages/blueman/gui/manager/ManagerDeviceList.py:350)
animating down 
_________
update (/usr/lib64/python2.7/site-packages/blueman/gui/DeviceList.py:179)
stopping monitor (not connected) 
_________
on_device_property_changed (/usr/lib64/python2.7/site-packages/blueman/gui/DeviceList.py:156)
list: device_prop_ch Connected 1 /org/bluez/hci0/dev_20_73_AB_62_9D_B3 () {} 
_________
row_update_event (/usr/lib64/python2.7/site-packages/blueman/gui/manager/ManagerDeviceList.py:289)
row update event Connected 1 
_________
init_services (/usr/lib64/python2.7/site-packages/blueman/main/Device.py:73)
Loading services 
_________
DisplayPinCode (/usr/lib64/python2.7/site-packages/blueman/bluez/Agent.py:67)
DisplayPinCode (/org/bluez/hci0/dev_20_73_AB_62_9D_B3, 552803) 
_________
Generate (/usr/lib64/python2.7/site-packages/blueman/gui/manager/ManagerDeviceMenu.py:213)
Bluetooth V3.0 TouchPad Keyboard 
_________
monitor_power_levels (/usr/lib64/python2.7/site-packages/blueman/gui/DeviceList.py:178)
starting monitor 
_________
level_setup_event (/usr/lib64/python2.7/site-packages/blueman/gui/manager/ManagerDeviceList.py:350)
animating up 
_________
on_device_property_changed (/usr/lib64/python2.7/site-packages/blueman/gui/DeviceList.py:156)
list: device_prop_ch Modalias bluetooth:v05ACp0239d011B /org/bluez/hci0/dev_20_73_AB_62_9D_B3 () {} 
_________
row_update_event (/usr/lib64/python2.7/site-packages/blueman/gui/manager/ManagerDeviceList.py:289)
row update event Modalias bluetooth:v05ACp0239d011B 
_________
init_services (/usr/lib64/python2.7/site-packages/blueman/main/Device.py:73)
Loading services 
_________
on_device_property_changed (/usr/lib64/python2.7/site-packages/blueman/gui/DeviceList.py:156)
list: device_prop_ch Paired 1 /org/bluez/hci0/dev_20_73_AB_62_9D_B3 () {} 
_________
row_update_event (/usr/lib64/python2.7/site-packages/blueman/gui/manager/ManagerDeviceList.py:289)
row update event Paired 1 
_________
Generate (/usr/lib64/python2.7/site-packages/blueman/gui/manager/ManagerDeviceMenu.py:213)
Bluetooth V3.0 TouchPad Keyboard 
_________
on_device_property_changed (/usr/lib64/python2.7/site-packages/blueman/gui/DeviceList.py:156)
list: device_prop_ch UUIDs dbus.Array([dbus.String(u'00001000-0000-1000-8000-00805f9b34fb'), dbus.String(u'00001124-0000-1000-8000-00805f9b34fb'), dbus.String(u'00001200-0000-1000-8000-00805f9b34fb')], signature=dbus.Signature('s'), variant_level=1) /org/bluez/hci0/dev_20_73_AB_62_9D_B3 () {} 
_________
row_update_event (/usr/lib64/python2.7/site-packages/blueman/gui/manager/ManagerDeviceList.py:289)
row update event UUIDs dbus.Array([dbus.String(u'00001000-0000-1000-8000-00805f9b34fb'), dbus.String(u'00001124-0000-1000-8000-00805f9b34fb'), dbus.String(u'00001200-0000-1000-8000-00805f9b34fb')], signature=dbus.Signature('s'), variant_level=1) 
_________
Generate (/usr/lib64/python2.7/site-packages/blueman/gui/manager/ManagerDeviceMenu.py:213)
Bluetooth V3.0 TouchPad Keyboard 
_________
on_device_property_changed (/usr/lib64/python2.7/site-packages/blueman/gui/DeviceList.py:156)
list: device_prop_ch Connected 0 /org/bluez/hci0/dev_20_73_AB_62_9D_B3 () {} 
_________
__init__ (/usr/lib64/python2.7/site-packages/blueman/main/Device.py:16)
caching initial properties 
_________
row_update_event (/usr/lib64/python2.7/site-packages/blueman/gui/manager/ManagerDeviceList.py:289)
row update event Connected 0 
_________
__del__ (/usr/lib64/python2.7/site-packages/blueman/main/Device.py:60)
deleting device /org/bluez/hci0/dev_20_73_AB_62_9D_B3 
_________
Destroy (/usr/lib64/python2.7/site-packages/blueman/main/Device.py:98)
invalidating device /org/bluez/hci0/dev_20_73_AB_62_9D_B3 
_________
Generate (/usr/lib64/python2.7/site-packages/blueman/gui/manager/ManagerDeviceMenu.py:213)
Bluetooth V3.0 TouchPad Keyboard 
_________
level_setup_event (/usr/lib64/python2.7/site-packages/blueman/gui/manager/ManagerDeviceList.py:350)
animating down 
_________
Generate (/usr/lib64/python2.7/site-packages/blueman/gui/manager/ManagerDeviceMenu.py:213)
Bluetooth V3.0 TouchPad Keyboard 
_________
Generate (/usr/lib64/python2.7/site-packages/blueman/gui/manager/ManagerDeviceMenu.py:213)
Bluetooth V3.0 TouchPad Keyboard 
_________
update (/usr/lib64/python2.7/site-packages/blueman/gui/DeviceList.py:179)
stopping monitor (not connected) 
_________
on_device_property_changed (/usr/lib64/python2.7/site-packages/blueman/gui/DeviceList.py:156)
list: device_prop_ch Connected 1 /org/bluez/hci0/dev_20_73_AB_62_9D_B3 () {} 
_________
row_update_event (/usr/lib64/python2.7/site-packages/blueman/gui/manager/ManagerDeviceList.py:289)
row update event Connected 1 
_________
Generate (/usr/lib64/python2.7/site-packages/blueman/gui/manager/ManagerDeviceMenu.py:213)
Bluetooth V3.0 TouchPad Keyboard 
_________
monitor_power_levels (/usr/lib64/python2.7/site-packages/blueman/gui/DeviceList.py:178)
starting monitor 
_________
Authorize (/usr/lib64/python2.7/site-packages/blueman/main/applet/BluezAgent.py:209)
Agent.Authorize 
_________
level_setup_event (/usr/lib64/python2.7/site-packages/blueman/gui/manager/ManagerDeviceList.py:350)
animating up 
_________
on_device_property_changed (/usr/lib64/python2.7/site-packages/blueman/gui/DeviceList.py:156)
list: device_prop_ch Class 1344 /org/bluez/hci0/dev_20_73_AB_62_9D_B3 () {} 
_________
row_update_event (/usr/lib64/python2.7/site-packages/blueman/gui/manager/ManagerDeviceList.py:289)
row update event Class 1344 
ERROR:dbus.connection:Exception in handler for D-Bus signal:
Traceback (most recent call last):
  File "/usr/lib64/python2.7/site-packages/dbus/connection.py", line 230, in maybe_handle_message
    self._handler(*args, **kwargs)
  File "/usr/lib64/python2.7/site-packages/blueman/bluez/PropertiesBlueZInterface.py", line 51, in wrapper
    handler(name, value, **kwargs)
  File "/usr/lib64/python2.7/site-packages/blueman/gui/DeviceList.py", line 163, in on_device_property_changed
    self.row_update_event(iter, key, value)
  File "/usr/lib64/python2.7/site-packages/blueman/gui/manager/ManagerDeviceList.py", line 346, in row_update_event
    c = self.make_caption(value, self.get_device_class(device), device.Address)
  File "/usr/lib64/python2.7/site-packages/blueman/gui/manager/ManagerDeviceList.py", line 230, in make_caption
    return "<span size='x-large'>%(0)s</span>\n<span size='small'>%(1)s</span>\n<i>%(2)s</i>" % {"0":cgi.escape(name), "1":klass.capitalize(), "2":address}
  File "/usr/lib64/python2.7/cgi.py", line 1043, in escape
    s = s.replace("&", "&amp;") # Must be done first!
AttributeError: 'dbus.UInt32' object has no attribute 'replace'
_________
Generate (/usr/lib64/python2.7/site-packages/blueman/gui/manager/ManagerDeviceMenu.py:213)
Bluetooth V3.0 TouchPad Keyboard 
_________
Generate (/usr/lib64/python2.7/site-packages/blueman/gui/manager/ManagerDeviceMenu.py:213)
Bluetooth V3.0 TouchPad Keyboard 
_________
Generate (/usr/lib64/python2.7/site-packages/blueman/gui/manager/ManagerDeviceMenu.py:213)
Bluetooth V3.0 TouchPad Keyboard 
_________
Generate (/usr/lib64/python2.7/site-packages/blueman/gui/manager/ManagerDeviceMenu.py:213)
Bluetooth V3.0 TouchPad Keyboard 
_________
Generate (/usr/lib64/python2.7/site-packages/blueman/gui/manager/ManagerDeviceMenu.py:213)
Bluetooth V3.0 TouchPad Keyboard 
_________
on_auth_action (/usr/lib64/python2.7/site-packages/blueman/main/applet/BluezAgent.py:212)
always 
_________
on_device_property_changed (/usr/lib64/python2.7/site-packages/blueman/gui/DeviceList.py:156)
list: device_prop_ch Trusted 1 /org/bluez/hci0/dev_20_73_AB_62_9D_B3 () {} 
_________
row_update_event (/usr/lib64/python2.7/site-packages/blueman/gui/manager/ManagerDeviceList.py:289)
row update event Trusted 1 
_________
Generate (/usr/lib64/python2.7/site-packages/blueman/gui/manager/ManagerDeviceMenu.py:213)
Bluetooth V3.0 TouchPad Keyboard 
```

**Comment by cschramm** [comment]
Thanks for your quick feedback. I've added a small patch that should fix the exception and will implement a notification for the pin code / passkey in the next couple of days I hope.

**Comment by cschramm** [comment]
OK, hacked it together; you can test the branch.

**Comment by patrakov** [comment]
Tested. It indeed shows a notification, and it may be useful under a desktop environment that never erases notifications from screen. However, under Mate desktop, the notification disappears from the screen too quickly for me to retype the code.

**Comment by patrakov** [comment]
If I type the code correctly, the bluetooth symbol on the icon turns green for a few seconds and then turns white again.

**Comment by patrakov** [comment]
Actually, the notification disappears when I start typing the code. So the workflow "retype 3 digits, look again, retype 3 other digits" doesn't work, while "remember and retype all 6 digits" works (but it is a kind of unexpected memory torture)

**Comment by infirit** [comment]
Maybe slightly off-topic but relevant, perhaps it is an idea to use a database of "known" pincodes for certain devices like [gnome-bluetooth](https://git.gnome.org/browse/gnome-bluetooth/tree/lib/pin-code-database.xml).

**Comment by cschramm** [comment]
I like memory torture. No, seriously, I'll try to set an infinite timeout. If that doesn't work, I'll use a custom dialog.

> If I type the code correctly, the bluetooth symbol on the icon turns green for a few seconds and then turns white again.

Does the keyboard work during and after that?

@infirit Interesting, thanks for mentioning this. It's more related to #117 though. In this case, if I got it correctly, the PC chooses the passkey / pin code and the user needs to enter it on the keyboard.

**Comment by cschramm** [comment]
@patrakov Please give the notification another try.

**Comment by patrakov** [comment]
It now shows a notification with a PIN code that I can retype. After that, it disconnects the keyboard, but shows it as paired. After a 20-second delay, a HID device authorization prompt appears, and, after I authorize the keyboard, it works. So the question now is "why the disconnection, delay and the extra authorization dialog?". Didn't I, by entering the PIN code, allow the keyboard to be used?

Here is the log:

```
aep@aep-haswell ~ $ blueman-applet
Loading configuration plugins

(blueman-applet:9841): Gtk-WARNING **: Theme parsing error: gtk-widgets.css:2:16: Theming engine 'adwaita' not found
blueman-applet version 1.99.alpha1 starting
Stale PID, overwriting
Using file config backend
_________
Load (/usr/lib64/python2.7/site-packages/blueman/main/PluginManager.py:54)
['AppIndicator', 'ExitItem', 'ShowConnected', 'AuthAgent', 'DhcpClient', 'NMPANSupport', 'DBusService', 'DiscvManager', 'NetUsage', 'Headset', 'KillSwitch', 'Menu', 'Networking', 'PowerManager', 'PPPSupport', 'NMDUNSupport', 'RecentConns', 'SerialManager', 'StandardItems', 'StatusIcon', 'TransferService'] 
_________
Load (/usr/lib64/python2.7/site-packages/blueman/main/PluginManager.py:54)
Unable to load plugin module AppIndicator
cannot import name AppIndicator3 
Using file config backend
_________
get_interface_version (/usr/lib64/python2.7/site-packages/blueman/bluez/BlueZInterface.py:8)
Detected BlueZ 5 
_________
__load_plugin (/usr/lib64/python2.7/site-packages/blueman/main/PluginManager.py:127)
loading <class 'blueman.plugins.applet.StatusIcon.StatusIcon'> 
_________
__load_plugin (/usr/lib64/python2.7/site-packages/blueman/main/PluginManager.py:127)
loading <class 'blueman.plugins.applet.DBusService.DBusService'> 
_________
__load_plugin (/usr/lib64/python2.7/site-packages/blueman/main/PluginManager.py:127)
loading <class 'blueman.plugins.applet.Headset.Headset'> 
Using file config backend
_________
__load_plugin (/usr/lib64/python2.7/site-packages/blueman/main/PluginManager.py:127)
loading <class 'blueman.plugins.applet.NMDUNSupport.NMDUNSupport'> 
_________
__load_plugin (/usr/lib64/python2.7/site-packages/blueman/main/PluginManager.py:127)
loading <class 'blueman.plugins.applet.SerialManager.SerialManager'> 
Using file config backend
_________
__load_plugin (/usr/lib64/python2.7/site-packages/blueman/main/PluginManager.py:127)
loading <class 'blueman.plugins.applet.Menu.Menu'> 
_________
__load_plugin (/usr/lib64/python2.7/site-packages/blueman/main/PluginManager.py:127)
loading <class 'blueman.plugins.applet.ExitItem.ExitItem'> 
_________
__load_plugin (/usr/lib64/python2.7/site-packages/blueman/main/PluginManager.py:127)
loading <class 'blueman.plugins.applet.AuthAgent.AuthAgent'> 
_________
__load_plugin (/usr/lib64/python2.7/site-packages/blueman/main/PluginManager.py:127)
loading <class 'blueman.plugins.applet.ShowConnected.ShowConnected'> 
_________
__load_plugin (/usr/lib64/python2.7/site-packages/blueman/main/PluginManager.py:127)
Not loading DhcpClient because it's conflict has higher priority 
_________
__load_plugin (/usr/lib64/python2.7/site-packages/blueman/main/PluginManager.py:127)
loading <class 'blueman.plugins.applet.TransferService.TransferService'> 
_________
__load_plugin (/usr/lib64/python2.7/site-packages/blueman/main/PluginManager.py:127)
loading <class 'blueman.plugins.applet.DiscvManager.DiscvManager'> 
Using file config backend
_________
update_menuitems (/usr/lib64/python2.7/site-packages/blueman/plugins/applet/DiscvManager.py:114)
warning: Adapter is None 
_________
__load_plugin (/usr/lib64/python2.7/site-packages/blueman/main/PluginManager.py:127)
loading <class 'blueman.plugins.applet.RecentConns.RecentConns'> 
Using file config backend
_________
__load_plugin (/usr/lib64/python2.7/site-packages/blueman/main/PluginManager.py:127)
Not loading PPPSupport because it's conflict has higher priority 
_________
__load_plugin (/usr/lib64/python2.7/site-packages/blueman/main/PluginManager.py:127)
loading <class 'blueman.plugins.applet.StandardItems.StandardItems'> 
_________
__load_plugin (/usr/lib64/python2.7/site-packages/blueman/main/PluginManager.py:127)
loading <class 'blueman.plugins.applet.PowerManager.PowerManager'> 
_________
__load_plugin (/usr/lib64/python2.7/site-packages/blueman/main/PluginManager.py:127)
loading <class 'blueman.plugins.applet.NMPANSupport.NMPANSupport'> 
_________
on_obex_owner_changed (/usr/lib64/python2.7/site-packages/blueman/plugins/applet/TransferService.py:70)
obex owner changed: :1.37 
Using file config backend
_________
on_started (/usr/lib64/python2.7/site-packages/blueman/main/applet/Transfer.py:72)
opp Started 
_________
on_started (/usr/lib64/python2.7/site-packages/blueman/main/applet/Transfer.py:72)
ftp Started 
_________
on_dbus_name_owner_change (/usr/bin/blueman-applet:86)
org.bluez owner changed to :1.14 
_________
register_agent (/usr/lib64/python2.7/site-packages/blueman/plugins/applet/AuthAgent.py:57)
Registering agent 
_________
__del__ (/usr/lib64/python2.7/site-packages/blueman/main/Device.py:60)
deleting device None 
_________
Destroy (/usr/lib64/python2.7/site-packages/blueman/main/Device.py:98)
invalidating device None 
Exception AttributeError: "'NoneType' object has no attribute 'Signals'" in <bound method Device.__del__ of <Device object at 0x7fe681d57730 (blueman+main+Device+Device at 0x254ec20)>> ignored
_________
initialize (/usr/lib64/python2.7/site-packages/blueman/plugins/applet/RecentConns.py:138)
rebuilding menu 
_________
__del__ (/usr/lib64/python2.7/site-packages/blueman/main/Device.py:60)
deleting device None 
_________
Destroy (/usr/lib64/python2.7/site-packages/blueman/main/Device.py:98)
invalidating device None 
Exception AttributeError: "'NoneType' object has no attribute 'Signals'" in <bound method Device.__del__ of <Device object at 0x7fe681d57780 (blueman+main+Device+Device at 0x251a160)>> ignored
_________
enumerate_connections (/usr/lib64/python2.7/site-packages/blueman/plugins/applet/ShowConnected.py:45)
Found 0 existing connections 
Loading configuration plugins
Using file config backend

(blueman-manager:9845): Gtk-WARNING **: Theme parsing error: gtk-widgets.css:2:16: Theming engine 'adwaita' not found
_________
Load (/usr/lib64/python2.7/site-packages/blueman/main/PluginManager.py:54)
['PulseAudioProfile', 'Services'] 
_________
__load_plugin (/usr/lib64/python2.7/site-packages/blueman/main/PluginManager.py:127)
loading <class 'blueman.plugins.manager.Services.Services'> 
_________
__load_plugin (/usr/lib64/python2.7/site-packages/blueman/main/PluginManager.py:127)
loading <class 'blueman.plugins.manager.PulseAudioProfile.PulseAudioProfile'> 
_________
pa_context_event (/usr/lib64/python2.7/site-packages/blueman/main/PulseAudioUtils.py:306)
1 
blueman-manager version 1.99.alpha1 starting
_________
pa_context_event (/usr/lib64/python2.7/site-packages/blueman/main/PulseAudioUtils.py:306)
2 
_________
on_bluez_name_owner_changed (/usr/bin/blueman-manager:105)
org.bluez owner changed to  :1.14 
Using file config backend
_________
get_interface_version (/usr/lib64/python2.7/site-packages/blueman/bluez/BlueZInterface.py:8)
Detected BlueZ 5 
_________
SetAdapter (/usr/lib64/python2.7/site-packages/blueman/gui/DeviceList.py:264)
None 
_________
on_adapter_changed (/usr/lib64/python2.7/site-packages/blueman/gui/manager/ManagerToolbar.py:74)
toolbar adapter /org/bluez/hci0 
_________
pa_context_event (/usr/lib64/python2.7/site-packages/blueman/main/PulseAudioUtils.py:306)
3 
_________
pa_context_event (/usr/lib64/python2.7/site-packages/blueman/main/PulseAudioUtils.py:306)
4 
_________
on_pa_ready (/usr/lib64/python2.7/site-packages/blueman/plugins/manager/PulseAudioProfile.py:24)
connected 
_________
<lambda> (/usr/lib64/python2.7/site-packages/blueman/main/PulseAudioUtils.py:318)
1 
_________
on_property_changed (/usr/lib64/python2.7/site-packages/blueman/gui/DeviceList.py:148)
adapter propery changed Discovering 1 
_________
on_adapter_property_changed (/usr/lib64/python2.7/site-packages/blueman/plugins/applet/DiscvManager.py:87)
prop Discovering 1 
_________
on_device_created (/usr/lib64/python2.7/site-packages/blueman/gui/DeviceList.py:246)
created /org/bluez/hci0/dev_20_73_AB_62_9D_B3 
_________
__init__ (/usr/lib64/python2.7/site-packages/blueman/main/Device.py:16)
caching initial properties 
_________
add_device (/usr/lib64/python2.7/site-packages/blueman/gui/DeviceList.py:310)
adding new device 
_________
do_cache (/usr/lib64/python2.7/site-packages/blueman/gui/DeviceList.py:489)
Caching new device 20:73:AB:62:9D:B3 
_________
row_update_event (/usr/lib64/python2.7/site-packages/blueman/gui/manager/ManagerDeviceList.py:289)
row update event Fake False 
_________
row_update_event (/usr/lib64/python2.7/site-packages/blueman/gui/manager/ManagerDeviceList.py:289)
row update event Trusted 0 
_________
row_update_event (/usr/lib64/python2.7/site-packages/blueman/gui/manager/ManagerDeviceList.py:289)
row update event Paired 0 
_________
on_device_property_changed (/usr/lib64/python2.7/site-packages/blueman/gui/DeviceList.py:156)
list: device_prop_ch RSSI -56 /org/bluez/hci0/dev_20_73_AB_62_9D_B3 () {} 
_________
row_update_event (/usr/lib64/python2.7/site-packages/blueman/gui/manager/ManagerDeviceList.py:289)
row update event RSSI -56 
opacity 222
_________
on_property_changed (/usr/lib64/python2.7/site-packages/blueman/gui/DeviceList.py:148)
adapter propery changed Discovering 0 
_________
on_adapter_property_changed (/usr/lib64/python2.7/site-packages/blueman/plugins/applet/DiscvManager.py:87)
prop Discovering 0 
_________
init_services (/usr/lib64/python2.7/site-packages/blueman/main/Device.py:73)
Loading services 
_________
Generate (/usr/lib64/python2.7/site-packages/blueman/gui/manager/ManagerDeviceMenu.py:213)
Bluetooth V3.0 TouchPad Keyboard 
_________
on_device_property_changed (/usr/lib64/python2.7/site-packages/blueman/gui/DeviceList.py:156)
list: device_prop_ch Connected 1 /org/bluez/hci0/dev_20_73_AB_62_9D_B3 () {} 
_________
row_update_event (/usr/lib64/python2.7/site-packages/blueman/gui/manager/ManagerDeviceList.py:289)
row update event Connected 1 
_________
DisplayPinCode (/usr/lib64/python2.7/site-packages/blueman/main/applet/BluezAgent.py:176)
DisplayPinCode (/org/bluez/hci0/dev_20_73_AB_62_9D_B3, 925201) 
_________
init_services (/usr/lib64/python2.7/site-packages/blueman/main/Device.py:73)
Loading services 
_________
Generate (/usr/lib64/python2.7/site-packages/blueman/gui/manager/ManagerDeviceMenu.py:213)
Bluetooth V3.0 TouchPad Keyboard 
_________
monitor_power_levels (/usr/lib64/python2.7/site-packages/blueman/gui/DeviceList.py:178)
starting monitor 
_________
level_setup_event (/usr/lib64/python2.7/site-packages/blueman/gui/manager/ManagerDeviceList.py:350)
animating up 
_________
on_device_property_changed (/usr/lib64/python2.7/site-packages/blueman/gui/DeviceList.py:156)
list: device_prop_ch Modalias bluetooth:v05ACp0239d011B /org/bluez/hci0/dev_20_73_AB_62_9D_B3 () {} 
_________
row_update_event (/usr/lib64/python2.7/site-packages/blueman/gui/manager/ManagerDeviceList.py:289)
row update event Modalias bluetooth:v05ACp0239d011B 
_________
init_services (/usr/lib64/python2.7/site-packages/blueman/main/Device.py:73)
Loading services 
_________
on_device_property_changed (/usr/lib64/python2.7/site-packages/blueman/gui/DeviceList.py:156)
list: device_prop_ch Paired 1 /org/bluez/hci0/dev_20_73_AB_62_9D_B3 () {} 
_________
row_update_event (/usr/lib64/python2.7/site-packages/blueman/gui/manager/ManagerDeviceList.py:289)
row update event Paired 1 
_________
Generate (/usr/lib64/python2.7/site-packages/blueman/gui/manager/ManagerDeviceMenu.py:213)
Bluetooth V3.0 TouchPad Keyboard 
_________
on_device_property_changed (/usr/lib64/python2.7/site-packages/blueman/gui/DeviceList.py:156)
list: device_prop_ch UUIDs dbus.Array([dbus.String(u'00001000-0000-1000-8000-00805f9b34fb'), dbus.String(u'00001124-0000-1000-8000-00805f9b34fb'), dbus.String(u'00001200-0000-1000-8000-00805f9b34fb')], signature=dbus.Signature('s'), variant_level=1) /org/bluez/hci0/dev_20_73_AB_62_9D_B3 () {} 
_________
row_update_event (/usr/lib64/python2.7/site-packages/blueman/gui/manager/ManagerDeviceList.py:289)
row update event UUIDs dbus.Array([dbus.String(u'00001000-0000-1000-8000-00805f9b34fb'), dbus.String(u'00001124-0000-1000-8000-00805f9b34fb'), dbus.String(u'00001200-0000-1000-8000-00805f9b34fb')], signature=dbus.Signature('s'), variant_level=1) 
_________
Generate (/usr/lib64/python2.7/site-packages/blueman/gui/manager/ManagerDeviceMenu.py:213)
Bluetooth V3.0 TouchPad Keyboard 
_________
on_device_property_changed (/usr/lib64/python2.7/site-packages/blueman/gui/DeviceList.py:156)
list: device_prop_ch Connected 0 /org/bluez/hci0/dev_20_73_AB_62_9D_B3 () {} 
_________
__init__ (/usr/lib64/python2.7/site-packages/blueman/main/Device.py:16)
caching initial properties 
_________
row_update_event (/usr/lib64/python2.7/site-packages/blueman/gui/manager/ManagerDeviceList.py:289)
row update event Connected 0 
_________
__del__ (/usr/lib64/python2.7/site-packages/blueman/main/Device.py:60)
deleting device /org/bluez/hci0/dev_20_73_AB_62_9D_B3 
_________
_________
Destroy (/usr/lib64/python2.7/site-packages/blueman/main/Device.py:98)
invalidating device /org/bluez/hci0/dev_20_73_AB_62_9D_B3 
Generate (/usr/lib64/python2.7/site-packages/blueman/gui/manager/ManagerDeviceMenu.py:213)
Bluetooth V3.0 TouchPad Keyboard 
_________
level_setup_event (/usr/lib64/python2.7/site-packages/blueman/gui/manager/ManagerDeviceList.py:350)
animating down 
_________
Generate (/usr/lib64/python2.7/site-packages/blueman/gui/manager/ManagerDeviceMenu.py:213)
Bluetooth V3.0 TouchPad Keyboard 
_________
Generate (/usr/lib64/python2.7/site-packages/blueman/gui/manager/ManagerDeviceMenu.py:213)
Bluetooth V3.0 TouchPad Keyboard 
_________
update (/usr/lib64/python2.7/site-packages/blueman/gui/DeviceList.py:179)
stopping monitor (not connected) 
_________
on_device_property_changed (/usr/lib64/python2.7/site-packages/blueman/gui/DeviceList.py:156)
list: device_prop_ch Connected 1 /org/bluez/hci0/dev_20_73_AB_62_9D_B3 () {} 
_________
row_update_event (/usr/lib64/python2.7/site-packages/blueman/gui/manager/ManagerDeviceList.py:289)
row update event Connected 1 
_________
Generate (/usr/lib64/python2.7/site-packages/blueman/gui/manager/ManagerDeviceMenu.py:213)
Bluetooth V3.0 TouchPad Keyboard 
_________
monitor_power_levels (/usr/lib64/python2.7/site-packages/blueman/gui/DeviceList.py:178)
starting monitor 
_________
Authorize (/usr/lib64/python2.7/site-packages/blueman/main/applet/BluezAgent.py:217)
Agent.Authorize 
_________
level_setup_event (/usr/lib64/python2.7/site-packages/blueman/gui/manager/ManagerDeviceList.py:350)
animating up 
_________
on_device_property_changed (/usr/lib64/python2.7/site-packages/blueman/gui/DeviceList.py:156)
list: device_prop_ch Class 1344 /org/bluez/hci0/dev_20_73_AB_62_9D_B3 () {} 
_________
row_update_event (/usr/lib64/python2.7/site-packages/blueman/gui/manager/ManagerDeviceList.py:289)
row update event Class 1344 
_________
on_device_property_changed (/usr/lib64/python2.7/site-packages/blueman/gui/DeviceList.py:156)
list: device_prop_ch Icon input-keyboard /org/bluez/hci0/dev_20_73_AB_62_9D_B3 () {} 
_________
row_update_event (/usr/lib64/python2.7/site-packages/blueman/gui/manager/ManagerDeviceList.py:289)
row update event Icon input-keyboard 
_________
Generate (/usr/lib64/python2.7/site-packages/blueman/gui/manager/ManagerDeviceMenu.py:213)
Bluetooth V3.0 TouchPad Keyboard 
_________
Generate (/usr/lib64/python2.7/site-packages/blueman/gui/manager/ManagerDeviceMenu.py:213)
Bluetooth V3.0 TouchPad Keyboard 
_________
Generate (/usr/lib64/python2.7/site-packages/blueman/gui/manager/ManagerDeviceMenu.py:213)
Bluetooth V3.0 TouchPad Keyboard 
_________
Generate (/usr/lib64/python2.7/site-packages/blueman/gui/manager/ManagerDeviceMenu.py:213)
Bluetooth V3.0 TouchPad Keyboard 
_________
Generate (/usr/lib64/python2.7/site-packages/blueman/gui/manager/ManagerDeviceMenu.py:213)
Bluetooth V3.0 TouchPad Keyboard 
_________
on_auth_action (/usr/lib64/python2.7/site-packages/blueman/main/applet/BluezAgent.py:220)
accept 
```

**Comment by patrakov** [comment]
Ugh, and now in the main window, the keyboard has a name "1344" instead of "Bluetooth V3.0 TouchPad Keyboard" :(

**Comment by infirit** [comment]
>@infirit Interesting, thanks for mentioning this. It's more related to #117 though. In this case, if I got it correctly, the PC chooses the passkey / pin code and the user needs to enter it on the keyboard.

The way I understand this is that gnome-bluetooth always pairs certain type of devices (device with no way to enter a pin on I think but do require a pun) without user input. It defaults for a headset to 0000 and sends this as pin when pairing, the headset is happy and pairs. However there are exceptions that require 1234 or some other combination and this xml file documents these exceptions in the top part. In other cases the normal pairing with random or no pin is used.

**Comment by cschramm** [comment]
> It now shows a notification with a PIN code that I can retype. After that, it disconnects the keyboard, but shows it as paired. After a 20-second delay, a HID device authorization prompt appears, and, after I authorize the keyboard, it works. So the question now is "why the disconnection, delay and the extra authorization dialog?". Didn't I, by entering the PIN code, allow the keyboard to be used?

I think that's the expected behavior. What you're doing with the PIN code is pairing the device. You should only need to do this once unless you un-pair on any side. Connecting to the service is a different and repeated step and requires explicit authorization. Sure, this is a little confusing for single service devices like a keyboard, but that's the way bluetooth works. In any case, you should be able to permanently allow the service by either responding with "Always" when asked or marking the device as trusted.

> Ugh, and now in the main window, the keyboard has a name "1344" instead of "Bluetooth V3.0 TouchPad Keyboard" :(

That's probably due to the fix of the exception you got: https://github.com/blueman-project/blueman/commit/287e9c6f5cb5e2c3e91e0ba3f45422f38fd40e00. It was actually doing good for you by avoiding that the device's Class property gets used for the caption. :wink: I have no idea why we're accepting both the Class and the Alias property in that code. We should probably only use Class if Alias is not available (I don't know if that could actually happen).

@infirit Yeap. But it's not related to this ticket. In case of the keyboard pairing discussed here, BlueZ chooses a key, blueman needs to display it and the user needs to enter it on the keyboard. In case of #117, it's the other way around: The device chooses the key and since it cannot be displayed, it's fixed and probably mentioned in the manual or something. BlueZ requests the key from blueman and currently we request it from the user (which does not work according to #117). We could use such a database instead of querying the user, but first the process needs to be working at all and according to #117 it doesn't. :wink:

**Comment by cschramm** [comment]
I removed the usage of the Class property in master (https://github.com/blueman-project/blueman/commit/525fad57022b97cea0e3722816001bc2fd6dacf9). It obviously never worked anyway due to the type error. Hope that brings you back the device alias.

I'll merge #162 tomorrow afternoon if no objections are raised and be kinda off-line for one week then.

**Comment by patrakov** [comment]
It shows "Bluetooth V3.0 TouchPad Keyboard". So, the only remaining cosmetic annoyance is that the PIN notification stays on screen even after I have entered the PIN code. But that's a separate bug, let's merge the changes and close this one.

**Comment by cschramm** [comment]
Yeah, I've had a thought about that already and it turns out, the agent class already has all it takes and I just need to store the notification object. Does https://github.com/blueman-project/blueman/commit/eabd0547c320d2d9fa6154d693c1e8c836244eb9 work as expected?

**Comment by patrakov** [comment]
No. It does not hide the notification.

Could you please paste some python code here so that I can test whether hiding notifications works at all in Mate?

**Comment by patrakov** [comment]
Uhm. That was because I typoed the PIN code. When I clodes the notification manually, it displayed another one, with a different code. That one did not go away, too, and resulted in a third notification, but I did not even attempt to enter the third PIN code, just dismissed the notification with a click. End result: the device shows up as paired, but "connect" fails with i/o error.

**Comment by cschramm** [comment]
> Could you please paste some python code here so that I can test whether hiding notifications works at all in Mate?

````python
from blueman.gui.Notification import Notification
n = NotificationBubble('Test', 'Test', 0)
n.close()
````

Should show a notification with infinite timeout after the second line and close it with the third one.

**Comment by patrakov** [comment]
With NotificationBubble changed to Notification, this works as expected. n.close() removes the notification and returns True.

**Comment by patrakov** [comment]
Still, PIN code notifications do not go away even if retyped correctly.

**Comment by cschramm** [comment]
Yeah, sorry, Notification of course. If that works, the workflow seems to be wrong. Do you see `Agent.Cancel` in the output of blueman-applet? That's what BlueZ [promises to call when the notification is no longer needed](https://git.kernel.org/cgit/bluetooth/bluez.git/tree/doc/agent-api.txt#n96). If it doesn't... Hrm... :unamused: 

The *Input/output error* might be the same I get with my mouse if I try to *re*connect to the HID service with BlueZ 5. I need to re-pair to get it working again. :point_right:  #13

**Comment by cschramm** [comment]
Btw. [this is where the notification should be closed](https://github.com/blueman-project/blueman/blob/eabd0547c320d2d9fa6154d693c1e8c836244eb9/blueman/main/applet/BluezAgent.py#L148) if it's stored in `self.n`.

**Comment by patrakov** [comment]
There is no Agent.Cancel in the output.
